phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -885,292 +867,215 @@ private static boolean clipKeyRangeBytes(RowKeySchema schema, int fieldIndex, in
-                if(newScan != null) {
+
+                Scan newScan = scanRanges.intersectScan(
+                        scan, keyRange.getLowerRange(), keyRange.getUpperRange(), keyOffset, crossesRegionBoundary);
+                if (newScan != null) {
                     ScanUtil.setLocalIndexAttributes(newScan, keyOffset, regionInfo.getStartKey(),
-                        regionInfo.getEndKey(), newScan.getStartRow(), newScan.getStopRow());
-                    // Boundary case of no GP in region after delaying adding of estimates
-                    if (!gpsInThisRegion && delayAddingEst) {
-                        updateEstimates(gps, guideIndex-1, estimates);
-                        gpsInThisRegion = true;
-                        delayAddingEst = false;
-                    }
-                } else if (!gpsInThisRegion) {
-                    delayAddingEst = false;
+                            regionInfo.getEndKey(), newScan.getStartRow(), newScan.getStopRow());
+                } else {
+                    logger.warn("Didn't generate scan for " + keyRange.toString());
                 }
-                scans = addNewScan(parallelScans, scans, newScan, endKey, true, regionLocation);
-                currentKeyBytes = endKey;
-                // We have a guide post in the region if the above loop was entered
-                // or if the current key is less than the region end key (since the loop
-                // may not have been entered if our scan end key is smaller than the
-                // first guide post in that region).
-                boolean gpsAfterStopKey = false;
-                gpsAvailableForAllRegions &=
-                    ( gpsInThisRegion && everNotDelayed) || // GP in this region
-                    ( regionIndex == startRegionIndex && gpsForFirstRegion ) || // GP in first region (before start key)
-                    ( gpsAfterStopKey = ( regionIndex == stopIndex && intersectWithGuidePosts && // GP in last region (after stop key)
-                            ( endRegionKey.length == 0 || // then check if gp is in the region
-                            currentGuidePost.compareTo(endRegionKey) < 0)  ) );
-                if (gpsAfterStopKey) {
-                    // If gp after stop key, but still in last region, track min ts as fallback
-                    fallbackTs =
-                            Math.min(fallbackTs,
-                                gps.getGuidePostTimestamps()[guideIndex]);
+
+                byte[] endKey;
+                if ((j < keyRangeCount - 1) || (i == countOfRegionsToScan - 1)) {
+                    endKey = keyRange.getUpperRange();
+                } else {
+                    endKey = regionInfo.getEndKey();
                 }
-                regionIndex++;
+
+                regionScans = addNewScan(parallelScans, regionScans, newScan, endKey, crossesRegionBoundary, regionLocation);
```

Comment:
Lets discuss the naming and usage a bit of this as in ParallelScanGrouper this is "startKey".  This is used for the crossesPrefixBoundary check and indicated in the interface as a startKey and compared to the last scans startRow Prefix looking for for example the salt byte changing or region changing.  By using end key is there any cases we might miss?

Reply:
Let double check this. This is exactly one of the places where we need to pay attention to. Hopefully I'll add test cases to cover the scenario.